File: dp832.cpp
#include <string.h>
#include <cassert>
#include "dp832.h"
#include <unistd.h>
using namespace std;



dp830::dp830(const char *devname) {
    fhandle_.open(devname,std::fstream::in | std::fstream::out | std::fstream::app);
}

dp830::~dp830(){fhandle_.close();}

int dp830::write_(const char *command) {
    assert(fhandle_);
    int len = strlen(command);
    fhandle_.write(command,len);
    //printf("write: -->%s<--\n",command);
    fhandle_.flush();
    return len;
}

int dp830::read_(char *buf, int len){
    assert(fhandle_);
    if (!buf) return 0;
    memset(buf,0,len);
    fhandle_.getline(buf,len);
    return fhandle_.gcount();
}

int dp830::GetState(int channel){
    char buf[40];
    char buf2[40];
    snprintf(buf,sizeof(buf),":OUTP:STAT? CH%d",channel);
    write_(buf);
    read_(buf2,sizeof(buf2));
    return strcmp(buf2,"OFF")?1:0;
}

int dp830::On(int channel) {
    char buf[40];
    snprintf(buf,sizeof(buf),":OUTP:STAT CH%d,ON",channel);
    return write_(buf);
}

int dp830::AllOn() {
    On(1);
    On(2);
    On(3);
    return 0;
}

int dp830::Off(int channel) {
    char buf[40];
    snprintf(buf,sizeof(buf),":OUTP CH%d,OFF",channel);
    return write_(buf);
}
int dp830::AllOff() {
    Off(1);
    Off(2);
    Off(3);
    return 0;
}


// Routines for setting operating parameters.  Note these are not
//  overvoltage/overcurrent protections.
int dp830::SetLimit_(int channel, const char *tag, double limit){
    char buf[40];
    snprintf(buf,sizeof(buf),":SOUR%d:%s:IMM %6.3f",channel,tag,limit);
    return write_(buf);
}
int dp830::SetVoltage(int channel, double limit){
    return SetLimit_(channel,"VOLT",limit);
}
int dp830::SetCurrent(int channel, double limit){
    return SetLimit_(channel,"CURR",limit);
}

double dp830::GetSetPoint_(int channel, const char *tag) {
    char buf[40];
    char buf2[40];
    snprintf(buf,sizeof(buf),":SOUR%d:%s?",channel,tag);
    write_(buf);
    read_(buf2,sizeof(buf2));
    return std::stod(buf2,NULL);
}
double dp830::GetVoltageSetPoint(int channel) {
    return GetSetPoint_(channel, "VOLT");
}
double dp830::GetCurrentSetPoint(int channel) {
    return GetSetPoint_(channel, "CURR");
}
// Measurement routines
double dp830::Measure_(int channel, const char *tag) {
    char buf[40];
    char buf2[40];
    snprintf(buf,sizeof(buf),":MEAS:%s? CH%d",tag,channel);
    write_(buf);
    read_(buf2,sizeof(buf2));
    return std::stod(buf2,NULL);
}
double dp830::MeasureVoltage(int channel) {
    return Measure_(channel,"VOLT");
}
double dp830::MeasureCurrent(int channel) {
    return Measure_(channel,"CURR");
}
double dp830::MeasurePower(int channel) {
    return Measure_(channel,"POWE");
}

//Bounce will set the output to 0.0V for delay seconds
//  Set to 0.0 instead of turning the output off so that
//    it will present a low impendence and help discharge
//    the target (if the circuit is designed in such a manner
//    to make this useful)
int dp830::Bounce(int channel, double delay) {
    double volt = GetVoltageSetPoint(channel);
    SetVoltage(channel, 0.0);
    usleep(delay*1e6);
    SetVoltage(channel, volt);
    return 0;
}

File: dp832.py
#!/usr/bin/env python
import time, socket
class dp832_mock:
  def __init__(self):
    pass
  def GetState(self, channel):
    return {"State":"OFF", "VLimit":10.2, "ILimit":2.2, "V":1.1, "I":2.2, "P":2.42}

class dp832:
  def __init__(self, fname="/dev/usbtmc3", ip=None, port=5555):
    self.ip = ip
    self.connected = False
    if self.ip:
      try:
        self.clientsock = socket.socket(family=socket.AF_INET, type=socket.SOCK_STREAM)
        self.clientsock.connect((self.ip, 5555))
        print("connection made to %s"%self.ip)
        self.connected = True
      except socket.error as msg:
        self.clientsock = None
        self.connected = False
    else:
      try:
        self.fd = open(fname,"w+", buffering=1)
        self.connected = True
      except IOError as e:
        self.fd = None
        print("Could not open %s -- %s"%(fname,e.strerror))
  def isConnected(self):
    return self.connected

  def Write(self, command):
    if self.ip:
      self.clientsock.send(command)
    else:
      self.fd.write(command)
  def Read(self):
    if self.ip:
      data = self.clientsock.recv(1024)
      return data
    else:
      return self.fd.readline()

  def GetState(self, channel):
    self.Write(":OUTP:STAT? CH%d\n"%channel)
    state = self.Read().strip()
    retval = {}
    retval["State"] = state
    retval["VLimit"] = self.GetVoltageSetPoint(channel)
    retval["ILimit"] = self.GetCurrentSetPoint(channel)
    retval["V"] = self.MeasureVoltage(channel)
    retval["I"] = self.MeasureCurrent(channel)
    retval["P"] = self.MeasurePower(channel)
    return retval

  def On(self, channel):
    self.Write(":OUTP:STAT CH%d,ON\n"%channel)
    self.GetState(channel)
  def Off(self, channel):
    self.Write(":OUTP:STAT CH%d,OFF\n"%channel)
    self.GetState(channel)
  def AllOn(self):
    for ch in range(1,4):
      self.On(ch)
  def AllOff(self):
    for ch in range(1,4):
      self.Off(ch)
  def Bounce(self, channel, delay=1.0):
    volt = self.GetVoltageSetPoint(channel)
    self.SetVoltage(channel, 0.0)
    time.sleep(delay)
    self.SetVoltage(channel, volt)


  def __SetLimit(self, channel, tag, limit):
    self.Write("SOUR%d:%s:IMM %6.3f\n"%(channel, tag, limit))
    self.GetState(channel)
  def SetVoltage(self, channel, voltage):
    self.__SetLimit(channel, "VOLT", voltage)
  def SetCurrent(self, channel, current):
    self.__SetLimit(channel, "CURR", current)


  def GetSetPoint(self, channel, tag):
    self.Write(":SOUR%d:%s?\n"%(channel, tag))
    return float(self.Read())
  def GetVoltageSetPoint(self, channel):
    return self.GetSetPoint(channel, "VOLT")
  def GetCurrentSetPoint(self, channel):
    return self.GetSetPoint(channel, "CURR")


  def Measure(self, channel, tag):
    self.Write(":MEAS:%s? CH%d\n"%(tag, channel))
    return float(self.Read())
  def MeasureVoltage(self, channel):
    return (self.Measure(channel, "VOLT"))
  def MeasureCurrent(self, channel):
    return (self.Measure(channel, "CURR"))
  def MeasurePower(self, channel):
    return (self.Measure(channel, "POWE"))
